plugin/src/main/kotlin/tanvd/grazi/language/LangDownloader.kt

```diff
@@ -0,0 +1,160 @@
+import tanvd.grazi.ide.ui.components.dsl.msg
+import java.io.File
+import java.net.URL
+import java.nio.file.Paths
+
+object LangDownloader {
+    private val logger = LoggerFactory.getLogger(LangDownloader::class.java)
+
+    private val repository: RepositorySystem
+    private val session: RepositorySystemSession
+    private val proxy = JreProxySelector()
+
+    private val MAVEN_CENTRAL_REPOSITORY =
+            RemoteRepository.Builder("central", "default", msg("grazi.maven.repo.url"))
+                    .setProxy(proxy.getProxy(msg("grazi.maven.repo.url"))).build()
+
+    init {
+        with (MavenRepositorySystemUtils.newServiceLocator()) {
+            addService(RepositoryConnectorFactory::class.java, BasicRepositoryConnectorFactory::class.java)
+            addService(TransporterFactory::class.java, FileTransporterFactory::class.java)
+            addService(TransporterFactory::class.java, HttpTransporterFactory::class.java)
+            setErrorHandler(object : DefaultServiceLocator.ErrorHandler() {
+                override fun serviceCreationFailed(type: Class<*>?, impl: Class<*>?, exception: Throwable?) {
+                    if (exception != null) {
+                        throw RuntimeException(exception)
+                    }
+                }
+            })
+
+            repository = getService(RepositorySystem::class.java)
+            session = MavenRepositorySystemUtils.newSession()
+            session.localRepositoryManager = repository.newLocalRepositoryManager(session, LocalRepository(File(GraziPlugin.installationFolder, "poms")))
+            session.proxySelector = JreProxySelector()
+            session.setReadOnly()
+        }
+    }
+
+    private fun Artifact.createDependency() = Dependency(this, JavaScopes.COMPILE, false, listOf(
+            Exclusion("org.languagetool", "languagetool-core", "", "jar"),
+            Exclusion("org.carrot2", "morfologik-fsa", "", "jar"),
+            Exclusion("org.carrot2", "morfologik-stemming", "", "jar"),
+            Exclusion("com.google.guava", "guava", "", "jar")
+    ))
+
+    private val Artifact.name
+        get() = "$artifactId-$version.jar"
+
+    private val Artifact.url
+        get() = "${MAVEN_CENTRAL_REPOSITORY.url}${groupId.replace(".", "/")}/$artifactId/$version/$name"
+
+    private fun DependencyNode.traverse(action: (DependencyNode) -> Unit) {
+        action(this)
+        this.children.forEach(action)
+    }
+
+    fun downloadMissingLanguages(project: Project?) {
+        val state = GraziConfig.get()
+
+        if (state.hasMissedLanguages()) {
+            state.enabledLanguages.filter { it.jLanguage == null }.forEach {
+                with(LangDownloader) { it.downloadLanguage(project) }
+            }
+
+            if (state.nativeLanguage.jLanguage == null) {
+                with(LangDownloader) { state.nativeLanguage.downloadLanguage(project) }
+            }
+
+            ProjectManager.getInstance().openProjects.forEach {
```

Comment:
Use notification for it

Reply:
Notification doesn't work because of state doesn't changed

Reply:
Hm, it may be a good point to move to other topic reset of DaemonCodeAnalyzer and send notification to it

Reply:
Are you sure we really need a whole topic for one implementation of 2 lines of code?

Reply:
I think we should be sure, that it is called asyncronously :)

Reply:
It's 100% asynchronous